HPCC-21186 Service level transforms should run for all methods

afishbeck opened the pull request:
We should treat service level transforms as common to all methods.
We can provide a way to selectively call common transforms in the
future.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12012, label: desdlAlwaysRunServiceLevelTransforms starts now. Estimated completion time is ~ 0.5 hour(s)

afishbeck commented:
@timothyklemm please review and if possible test this change.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: ffbdd9d14f4beb9319926ab2bd6ac0c29395fb28
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |808 |808 |0 |
|test (thor) |732 |732 |0 |
|test (roxie) |882 |882 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   17 sec (00:00:17) | 1471 sec (00:24:31) |   20 sec (00:00:20) | 1730 sec (00:28:50) |

timothyklemm reviewed (changes requested):
Preliminary testing was successful. I was able to move some method level transforms to service level. All transforms were processed, as evidenced by seeing the modified request in the log file.

All of the review comments are related to code redundancy. Not sure if old code is still necessary, but if it is it should share code with the new implementation and not duplicate it.

afishbeck committed:
HPCC-21186 Service level transforms should run for all methods

We should treat service level transforms as common to all methods.
We can provide a way to selectively call common transforms in the
future.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented on esp/services/esdl_svc_engine/esdl_svc_custom.hpp:
```diff
@@ -131,6 +131,8 @@ class CEsdlCustomTransformChoose : public CInterface
 interface IEsdlCustomTransform : extends IInterface
 {
        virtual void processTransform(IEspContext * context, IPropertyTree *tgtcfg, IPropertyTree *tgtctx, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, StringBuffer & request, IPropertyTree * bindingCfg)=0;
```
This method could come in handy, I'll keep it for now, but may decide to remove it later.

afishbeck commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -310,6 +310,12 @@ CEsdlCustomTransform::CEsdlCustomTransform(IPropertyTree &currentTransform)
 #endif
 }
 
+void CEsdlCustomTransform::processTransform(IEspContext * context, IPropertyTree *theroot, IXpathContext *xpathContext, const char *defaultTarget)
```
I've commoned up the code.

afishbeck commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -401,3 +407,100 @@ void CEsdlCustomTransform::processTransform(IEspContext * context, IPropertyTree
     }
 }
+
+void processServiceAndMethodTransforms(IEsdlCustomTransform *serviceCrt, IEsdlCustomTransform *methodCrt, IEspContext * context, IPropertyTree *tgtcfg, IPropertyTree *tgtctx, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, StringBuffer & request, IPropertyTree * bindingCfg)
```
I've commoned up the code and used an initializer_list.  May switch to a vector later if we need something more flexible.

afishbeck commented:
@timothyklemm updated and squashed please review.

HPCCSmoketest commented:
Process of PR-12012, label: desdlAlwaysRunServiceLevelTransforms starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3c9d2f7f23747005dc21a09c6d76ee8316c2692b
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |808 |808 |0 |
|test (thor) |732 |732 |0 |
|test (roxie) |882 |882 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  184 sec (00:03:04) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1497 sec (00:24:57) |   19 sec (00:00:19) | 1756 sec (00:29:16) |

timothyklemm reviewed (approved):
Reviewed and tested. Service and method level transforms both applied.
